[Lai Chok Hoe] iP

lchokhoe opened the pull request:
**DukeLCH**

DukeLCH is a useful aid that helps you remember things you have to do. It's,

- Text-based
- Easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/lchokhoe/ip)
2. double-click it
3. add your tasks
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:

- [x]  Managing Tasks
- [ ]  Managing Deadlines
- [ ]  Reminders (coming soon) 

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Duke {
    public static void main(String[] args) {
            new Duke("data\\duke.txt").run();
    }
}
```

lchokhoe committed:
test text file added

lchokhoe committed:
test text file deleted

lchokhoe committed:
Greet added to Duke

lchokhoe committed:
Commands class added

lchokhoe committed:
Renamed Duke to DukeLCH and changed default Greet response

lchokhoe committed:
Updated bye response

lchokhoe committed:
Added comments to methods in Commands Class and in Main Class

lchokhoe committed:
Added InputHistory class to track input history

lchokhoe committed:
Added cmdHIstory in Commands Class. echoInput method now adds input into cmdHistory with each call

lchokhoe committed:
Added list method to Commands Class that gets DukeLCH to list cmdHistory

lchokhoe committed:
Fixed a bug where count wasn't being incremented with each iteration in printAll() method

lchokhoe committed:
Updated echo message to correctly reflect DukeLCH's behaviour given a random user input

lchokhoe committed:
Added a new if-else statement to check for "list" command when reading user input

lchokhoe committed:
Changed the printAll() method in InputHistory class to return a String (from StringBuilder) instead to ensure "list" output matches demo output

lchokhoe committed:
Added Task Class

lchokhoe committed:
Made changes to names, added 2 methods to mark and unmark tasks,added 2 new keywords to look for in Duke.java

lchokhoe committed:
Fixed an issue with reading keyword with int inputs

lchokhoe committed:
Fixed a bug where the mark/unmark requests are offset by a value of 1

lchokhoe committed:
Updated the list() and unmark() methods in Commands Class to correctly print messages

lchokhoe committed:
Currently tackling Level-4

lchokhoe committed:
todo keyword added into DukeLCH

lchokhoe committed:
Removed default behaviour of adding type Task and addTask() methods. Only allowed to use "todo", "deadline" or "event" from this point onwards

lchokhoe committed:
deadline keyword added into DukeLCH

lchokhoe committed:
Added error message to "deadline" keyword and updated mark to work with new keywords

lchokhoe committed:
Updated unmark to work with new keywords

lchokhoe committed:
Removed exit() calls whenever the wrong input is detected to allow user to continue using DukeLCH

lchokhoe committed:
event keyword added into DukeLCH

lchokhoe committed:
Added Automated Text UI Testing

lchokhoe committed:
Added DukeException class

lchokhoe committed:
Added exception handling for keywords "mark" and "unmark"

lchokhoe committed:
Updated error message for keywords "mark" and "unmark"

lchokhoe committed:
Updated error handler of "mark: and "unmark" with new messages and to handle decimal entry inputs by user

lchokhoe committed:
Added exception handler for keywords "bye" and "list"

lchokhoe committed:
Updated exception handler for "mark" and "unmark"

lchokhoe committed:
exception handler added for keyword "deadline" and "todo"

lchokhoe committed:
exception handler added for keyword "event"

lchokhoe committed:
exception handler added when unrecognised keyword is detected

lchokhoe committed:
updated Commands Class to use ArrayList<>() instead of java Array to store records

lchokhoe committed:
added "delete" keyword to DukeLCH with exception handler (missing delete message)

lchokhoe committed:
added message print to "delete" keyword

lchokhoe committed:
Add JavaDoc comment to printAll()

This is for the clarity of the printAll() method implemented in the TaskHistory class

lchokhoe committed:
Adjust coding style for switch-case apllication in Duke.java

To adhere to the java coding standard set for this iP.

lchokhoe committed:
Duke class: Update Duke's response to "invalid keywords"

Duke's response to "invalid keywords" is to print a list of
recognized keywords that the user can use.

Updating the response will teach users the keywords they can use to
access the latest features of Duke.

Let's update the printed response to correctly reflect the current
list of recognized keywords.

lchokhoe committed:
Commands class: add startup() method

There does not exist a way for us to detect if ./data/duke.txt is a
valid file path that exists.

In order for us to implement our "save" feature in Duke, we require
./data/duke.txt to exist so we can update the duke.txt file after
each update by the user.

Let's implement a startup() method that is called at the beginning
of each session with Duke that checks for and creates the data
folder and duke.txt file.

This is so we can easily bug fix this startup() method in the future
in the case of any poor implementation.

The startup() method currently implemented is only able to check for
the existence of a folder named "data" in the root folder containing
Duke.java.

lchokhoe committed:
Commands class: update startup() method

startup() method is only capable of checking the existence of a
"data" directory in the root directory containing the Duke program.

This is only a means to determine if there is a need to manually
create the directory for future use.

Let's now provide an independent path to contain our duke.txt file
and also create the parent directory, child directory and duke.txt
file if it does not already exist.

startup() now creates a folder and duke.txt file that will be used
to keep track of a users task history.

The specific path used is "C:\Users\username\Duke\data". The main
method in Duke.java has also been updated to call startup() at the
beginning of each session.

lchokhoe committed:
Commands class: add restore() method

There does not exist a way for users to reuse a previously created
local copy of duke.txt.

This does not fulfil one of the features of Duke that we are trying
to implement. We want to be able to load the data from the hard disk
when Duke starts up

Let's now provide a way to load the contents of an existing duke.txt
file onto a session of Duke.

By creating a restore() method that takes in a File object as
argument, we are able to customize how Duke handles a custom
duke.txt script. This gives us the flexibility to format the
duke.txt file for our convenience.

To facilitate the implementation of restore(), we want to have
special constructors for the Todo, Deadline and Event classes that
can allow us to create tasks based on the information stored on the
duke.txt file.

Lastly, we also want to provide methods to add these tasks on the
duke.txt file in the background without interfering with the current
session of duke. This is done by overloading the add"" methods in
the Commands class

lchokhoe committed:
Commands class: add update() method

There does not exist a way for users to manually request Duke to
store the data of the current session onto their local device.

We want to provide the flexibility to our users to manually request
Duke to store the data of their current session. We also need a way
to trigger this storing of data when Duke is closed so this is a
step towards that.

Let's now provide a way to request Duke to update the local duke.txt
file with the latest records stored in Duke for the current session.

By creating a update() method, we are able to call for the update in
our code and also provide a way for the user to manually request for
this update without having to close Duke.

The ToDos, Deadline and Event classes have been provided methods to
help with update(). Also, a new keyword "update" is recognized by
Duke.

lchokhoe committed:
Duke class: update "bye" keyword behavior

Previously, we do not get Duke to update the local duke.txt file at
the end of a session.

If the user forgets to call "update" manually, the data created
or changed during the session is not saved.

Let's make it so that Duke will automatically save data to the user
hard disk by calling update() when "bye" keyword is used.

lchokhoe committed:
Duke class: add isDateValid() method

For keywords "deadline" and "event", Duke interprets the date & time
arguments provided as Strings and stores them word for word in it's
records.

We want to provide users with a more uniform response when it comes
to date & time such that any provided input will give the same
output.

Let's first introduce a method to validate date formats so we can
teach Duke different date formats and get it to recognize them.

We want to create a method that we can expand on in the future,
adding new date formats for Duke to recognize.

Currently, the method only has 2 date formats that it recognizes and
does not validate time.

lchokhoe committed:
Duke class: change isDateValid() method to
convertToDukeDate() method

isDateValid() returned a boolean that indicated if Duke recognized
the given date format or not.

This boolean needed to be handled separately outside of
isDateValid() before Duke will process the given date into a uniform
format.

Let's change the isDateValid() method into a method that immediately
does the conversion as well and returns a LocalDate object.

This gives as a more useful object to manipulate accordingly in our
other methods in the Commands class.

convertToDukeDate() has been moved from Duke.java to Commands.java
where it will be better used. Both "deadline" and "event" keywords
have had their implementations updated to include the use of
convertToDukeDate().

lchokhoe committed:
Duke class: add convertToDukeTime()

Duke currently treats time as a String and does not manipulate it in
any way.

We want Duke to be able to convert time inputs in 24-hour format to
a standardized output.

Let's implement a new method convertToDukeTime() that does this.

This gives us the flexibility to detect and convert 24-hour time
to 12-hour time for any commands that expect a time input from the
user.

"deadline" and "event" keywords have been updated to call
convertToDukeTime() when needed.

lchokhoe committed:
Duke class: update Duke to save state after every command

Duke currently only saves the state to the user's local hard drive
on request ("update") or at the end of a session.

This could led to potential lost of user data if an expected crash
leading to an unexpected end to the session without the user or
Duke invoking the update() command.

Let's make it so that Duke will automatically try to update the
local duke.txt file after every command.

We are simply adding a update() call after every keyword call.

lchokhoe committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Commands.java
#	src/main/java/TaskHistory.java

Resolution: keep both

lchokhoe committed:
Fix bugs from merging branch-Level-8

lchokhoe committed:
Add more OOP components (DukeUi, DukeStorage and DukeHistory)

lchokhoe committed:
Move related files to duke package

lchokhoe committed:
Add DukeTest.java for unit testing of Duke.

lchokhoe committed:
Add Gradle support with build.gradle

lchokhoe committed:
Add javaDocs comment for all methods in subclasses used by the main
program.

lchokhoe committed:
Adjust code to adhere to coding standard.

lchokhoe committed:
Duke package: add support for "find" keyword.

The only way for users to view the records in Duke during a session
is to use the "list" keyword.

This is inefficient and provides very little flexibility when it
comes to viewing records.

Let's add a feature that allows the user to view all tasks in Duke's
records according to a specified word or phrase.

We will implement a new keyword "find" that takes in a phrase, looks
through duke's history and returns a list of all tasks that has that
phrase in their description.

Support for "find" keyword has been added in subclasses supporting
the Duke package in the form of:
Commands class: add find() method
DukeHistory class: add findPhrase() method
DukeUi class: add printFoundNothing() method

jiale-c commented on src/main/java/duke/Commands.java:
```diff
@@ -0,0 +1,297 @@
+     *
+     * @param tokens A String[] of tokens inputted by the user.
+     * @param history An instance of DukeHistory.
+     */
```
I like that your javadoc is very clear and to the point :smiley:

jiale-c commented on src/main/java/duke/Deadlines.java:
```diff
@@ -0,0 +1,55 @@
+ * Represents Deadlines class that inherits from Task class.
+ */
+public class Deadlines extends Task {
+    private final String time;
```
I like that you used non public variables.

jiale-c commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,200 @@
+            String[] tokens = input.split("\\s");
+            String keyword = tokens[0];
+            switch (keyword) {
+            case "bye":
```
I like the use of switch here to run through the possibilities.

jiale-c reviewed (commented):
Overall code is very neat and the javadoc is very helpful to understand code.

whoisjunhong commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,200 @@
+                } catch (NumberFormatException | NullPointerException | IndexOutOfBoundsException ex) {
+                    ui.printInvalidArgumentError();
+                    break;
+                } catch (IOException e) {
+                    ui.printWriteError();
+                    break;
+                }
+            }
+            case "update": {
+                try {
+                    storage.update(history);
+                    break;
+                } catch (IOException e) {
+                    ui.printWriteError();
+                    break;
+                }
+            }
+            case "find": {
+                try {
+                    if (tokens.length <= 1) {
+                        throw new MissingFormatArgumentException("no argument detected");
+                    }
+                    cmd.find(tokens, history);
+                    break;
+                } catch (MissingFormatArgumentException ex) {
+                    ui.printMissingArgumentError(keyword);
+                    break;
+                } catch (DukeException e) {
+                    ui.printFoundNothing();
+                    break;
+                }
+            }
+            default:
+                try {
+                    throw new MissingFormatArgumentException("invalid keywords");
+                } catch (MissingFormatArgumentException ex) {
+                    ui.printListOfCommands();
+                    break;
+                }
+            }
```
I like how you managed to account for the different commands in `run()` using a `switch` statement, would it be better for `run()` to have a layer of abstraction such as a `Parser` to handle the different cases in the `switch` statement?

whoisjunhong commented on src/main/java/duke/DukeHistory.java:
```diff
@@ -0,0 +1,266 @@
+    public void addToDo(String description) {
+        ToDos tempToDo = new ToDos(description);
+        records.add(tempToDo);
+        String msg = "_______________________________________________________\n"
+                + "Understood, adding this task now:\n"
+                + "    " + tempToDo.getToDo()
+                + "Currently you have " + records.size() + " tasks in our records.\n"
+                + "_______________________________________________________\n";
+        System.out.println(msg);
```
Should this be abstracted into `DukeUi.java`? This applies to all `String msg` in this file. 😄

whoisjunhong commented on src/main/java/duke/Commands.java:
```diff
@@ -0,0 +1,297 @@
+            Date dummyDate1 = df1.parse(date);
+            LocalDate dukeDate1 = dummyDate1.toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
+            return dukeDate1.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+        } catch (ParseException e1) {
+            try {
+                Date dummyDate2 = df2.parse(date);
+                LocalDate dukeDate2 = LocalDate.parse(date);
+                return dukeDate2.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+            } catch (ParseException e2) {
+                return date;
+            }
+        }
+    }
+
+    /**
+     * A method that takes in a String representing a time period and checks if it's told in a 12-hour clock.
+     * If not, the method attempts to convert it into a 12-hour clock.
+     *
+     * @param time A String input that represents a time period.
+     * @return Either a default 12-hour clock format or the user input.
+     */
+
+    public String convertToDukeTime(String time) {
+        if (time.contains("am") || time.contains("pm")
+                || time.contains("AM") || time.contains("PM")
+                || time.contains("Am") || time.contains("Pm")) {
+            return time;
+        } else {
+            String splicedTime = time.substring(0, 4);
+            DateFormat inputFormat = new SimpleDateFormat("HHmm");
+            DateFormat outputFormat = new SimpleDateFormat("hh:mmaa");
+            try {
+                Date dukeTime = inputFormat.parse(splicedTime);
+                return outputFormat.format(dukeTime);
+            } catch (ParseException e) {
+                e.printStackTrace();
+                return time;
+            }
+        }
+    }
```
I like how to accounted for many scenarios to convert the date to a DukeDateTime. 👍 

However, should this be in `Commands.java`? Maybe it would be better to have a `Validation` class to simplifying the conversion and validation of a DukeDateTime.

whoisjunhong reviewed (commented):
Overall, code integrity and logic looks great! 🎉 

The common issue found is that many UI is not abstracted and would lead to many bugs and formatting issues when making a change in the future. 🤕 

Perhaps a more intuitive way of calling UI methods can be used throughout your Duke project? 😄

lchokhoe committed:
Use checkstyle to detect coding style violations.

lchokhoe committed:
Add gui support for DukeLCH using javaFX basics.

gigituang commented on src/main/java/Event.java:
```diff
@@ -8,16 +8,17 @@ public Event(String description, String date, String time) {
 
     public String getFormattedText() {
-        return "E:" + this.getMark() + ":" + this.getDescription() + ":" + this.timeFrame;
+        return "E>" + this.getMark() + ">" + this.getDescription() + ">" + this.date + ">" + this.time;
```
I think we can use this.type instead of hardcoding "E" for the event. This can be done for the other tasks too:)

gigituang commented on src/main/java/ToDos.java:
```diff
@@ -13,6 +13,6 @@ public String getToDo() {
 
     public String getFormattedText() {
-        return "T:" + this.getMark() + ":" + this.getDescription();
+        return "T>" + this.getMark() + ">" + this.getDescription();
```
Same here, you can use this .type.

gigituang commented on src/main/java/duke/Commands.java:
```diff
@@ -270,4 +270,28 @@ public void delete(int index, DukeHistory history)  {
+     * @param history An instance of DukeHistory.
+     * @throws DukeException If no task is found matching the inputted phrase.
+     */
+    public void find(String[] tokens, DukeHistory history) throws DukeException {
+        StringBuilder phrase = new StringBuilder();
+        for (int i = 1; i < tokens.length; i++) {
+            phrase.append(tokens[i]);
+            if (i != (tokens.length - 1)) {
+                phrase.append(" ");
+            }
+        }
+        String border = "_______________________________________________________\n";
+        System.out.println(border
+                + "Here is what we found:\n"
+                + history.findPhrase(phrase.toString()) + border);
+    }
```
This command class may be a little too long?
It would be great if you could split up the find and delete command, and extend them from a command class. It will be easier to sort and debug if you encounter any issues:)
an example could be `public class FindCommand extends Command {}`

gigituang commented on src/main/java/duke/Commands.java:
```diff
@@ -8,18 +8,34 @@
+     *
+     * @param date A String input that represents a date.
+     * @return Either a default formatted date or the user input.
+     */
```
Really love your detailed comments and how it clearly explains the use of each functions!:)

gigituang reviewed (commented):
Overall, I love how detailed your codes are as I can understand it pretty easily. Maybe one thing to note is to keep your classes a little shorter as it is pretty lengthy. You can consider splitting into smaller classes if that makes sense. Really good work! Keep it up!!

lchokhoe committed:
Add GUI support using javafx.

lchokhoe committed:
Revert "Add GUI support using javafx."

This reverts commit 7f52a249cd26df64518d83406842aea0d4715512.

lchokhoe committed:
Revert "Revert "Add GUI support using javafx.""

This reverts commit b960fde2a055f830b06d54096e400662b5953157.

lchokhoe committed:
Duke.java: assertions used so define assumptions so that the runtime
can verify them.

Currently, Duke is manually ensuring that all inputs have the right
number of tokens to ensure that the keywords used are provided the
appropriate amount of arguments.

This is inefficient and makes our code lengthier as a result.

We want to make use of java assert to verify these basic assumptions
rather then write specific if-else statements to handle this during
runtime.

Assert it a in-built function of Java that we can use to shorten our
code while still achieving the desired outcome.

As much as possible, all keywords are now asserting assumptions and
changes can be found within the run() method of Duke.java.

lchokhoe committed:
Update codeQuality branch.

Create byeCommand, listCommand, markCommand,
unmarkCommand, todoCommand and deadlineCommand classes.

lchokhoe committed:
Update codeQuality.

Add implementation to byeCommand, listCommand, markCommand,
unmarkCommand, todoCommand and deadlineCommand classes.

lchokhoe committed:
Update code Quality.
Add DeleteCommand, EventCommand, FindCommand and UpdateCommand.

Currently Command class, DukeHistory class and Duke class have been
sharing the work of dealing with all valid Duke Commands.

This has caused the overall code to be excessively long and
unreadable.

Let's update the code quality of Duke. Specifically, let's lessen
the burden on the above mentioned classes when handling valid
Commands. We will implement specific sub-Command classes that
inherit from the Commands class and offload all the code logic to
those classes. We will strive towards SLAP in our implementation.

This is so our code becomes shorter and much more readable.

The past 2 commits has added the other sub-Command classes. Also the
GUI has been fixed so that it now displays the full output without
truncating the String.

lchokhoe committed:
Merge pull request #3 from lchokhoe/branch-A-CodeQuality

Merge branch-A-CodeQuality.

lchokhoe committed:
Merge branch 'master' of https://github.com/lchokhoe/ip

* 'master' of https://github.com/lchokhoe/ip:
  Update code Quality. Add DeleteCommand, EventCommand, FindCommand and UpdateCommand.
  Update codeQuality.
  Update codeQuality branch.

lchokhoe committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/duke/Commands.java
#	src/main/java/duke/Duke.java
#	src/main/resources/view/DialogBox.fxml

lchokhoe committed:
Remove unused import statement from Commands.java.

lchokhoe committed:
Merge pull request #2 from lchokhoe/branch-A-Assertions

Merge branch-A-Assertions.

lchokhoe committed:
Merge branch 'master' of https://github.com/lchokhoe/ip

* 'master' of https://github.com/lchokhoe/ip:
  Remove unused import statement from Commands.java.
  Duke.java: assertions used so define assumptions so that the runtime can verify them.

lchokhoe committed:
Add D-Contacts extension and update javadocs for all classes.

lchokhoe committed:
Merge branch 'branch-D-Contacts'

lchokhoe committed:
Add a brief user guide.

lchokhoe committed:
Set theme jekyll-theme-minimal

lchokhoe committed:
Fix Junit test class.

lchokhoe committed:
Add greeting when DukeLCH is first launched.

lchokhoe committed:
Fix bye command such that DukeLCH application properly closes
itself when 'bye' is used.

lchokhoe committed:
Change duke and user images for cute animals images.

Add Ui.png into docs folder.

lchokhoe committed:
Merge branch 'master' of https://github.com/lchokhoe/ip

lchokhoe committed:
Rename Ui.png.png file to Ui.png.

lchokhoe committed:
Fix 'bye' command so that it now properly calls System.exit() when
used by the user.

lchokhoe committed:
Fix file separator bug that was causing DukeLCH to create 2
directories for Linux users.
